### PR TITLE
Extraction of storage_options from URL

### PR DESCRIPTION
**Pull request by manjuransari**
PR for issue #691

### CODE REVIEW

File: petastorm/fs_utils.py [hunk 1]
```diff
@@ -134,8 +131,12 @@ def __init__(self, dataset_url, hadoop_configuration=None, connector=HdfsConnect
 
             storage_options = storage_options or {}
             protocol = self._parsed_dataset_url.scheme
-            self._filesystem = fsspec.filesystem(protocol, **storage_options)
-            self._filesystem_factory = lambda: fsspec.filesystem(protocol, **storage_options)
+            cls = fsspec.get_filesystem_class(protocol)
+            options = cls._get_kwargs_from_urls(self._dataset_url)
```
Comment (selitvin) [inline petastorm/fs_utils.py:135]: Please don't use private methods of a third party library.

File: petastorm/fs_utils.py [hunk 2]
```diff
@@ -134,8 +131,12 @@ def __init__(self, dataset_url, hadoop_configuration=None, connector=HdfsConnect
...
+            update_storage_options(options, storage_options)
+            fs = cls(**options)
+            self._filesystem = cls(**options)
```
Comment (selitvin) [inline petastorm/fs_utils.py:138]: Why do we construct `cls` twice? (lines 138 and 137)

File: petastorm/fs_utils.py [hunk 3]
```diff
@@ -211,3 +212,15 @@ def normalize_dir_url(dataset_url):
     dataset_url = dataset_url[:-1] if dataset_url[-1] == '/' else dataset_url
     logger.debug('directory url: %s', dataset_url)
     return dataset_url
+
+def update_storage_options(options, inherited=None):
+    if not inherited:
```
Comment (selitvin) [inline petastorm/fs_utils.py:217]: Please add a docstring and a unit test.

**Review by selitvin — CHANGES_REQUESTED** [review]
Is it possible to add some unittests to cover new functionality?

File: petastorm/fs_utils.py (hunk 1 above)
Comment (manjuransari) [reply petastorm/fs_utils.py:135]: All third party libraries being used by fsspec implement class AbstractFileSystem. AbstractFileSystem class have this method which either developer can over ride according to his/her needs. More info: https://filesystem-spec.readthedocs.io/en/latest/developer.html

File: petastorm/fs_utils.py (hunk 2 above)
Comment (manjuransari) [reply petastorm/fs_utils.py:138]: typo. Will update the code.

File: petastorm/fs_utils.py [hunk 4]
```diff
@@ -17,23 +17,20 @@
...
-    if parsed_url.scheme.lower() in ['file', 'hdfs']:
-        return parsed_url.path
+def get_dataset_path(dataset_url):
+    """Return only path part of full URL, according to appropriate backend"""
+    if urlparse(dataset_url).scheme.lower() in ['file', 'hdfs']:
+        return urlparse(dataset_url).path
 
-    # s3-style filesystems requires the form `bucket/path`
-    return parsed_url.netloc + parsed_url.path
+    _url = strip_protocol(dataset_url)
```
[the remark this answers is not in the document]
Comment (manjuransari) [reply petastorm/fs_utils.py:32]: Its both clean up and change for handling the url that has storage options within them. All third party libraries being used by fsspec implement class AbstractFileSystem. AbstractFileSystem class have this method which either developer can over ride according to his/her needs. More info: https://filesystem-spec.readthedocs.io/en/latest/developer.html

File: petastorm/fs_utils.py (hunk 2 above)
Comment (manjuransari) [reply petastorm/fs_utils.py:138]: typo. Updated changes.

File: petastorm/fs_utils.py (hunk 3 above)
Comment (manjuransari) [reply petastorm/fs_utils.py]: Removed this method. Using the same method from fsspec library.

**Review by selitvin — COMMENTED** [review]
All looks good to me. One last ask: please update docs/release-notes.rst with the description of your change and then we'll land.

Thank you for your contribution!

**Review by selitvin — APPROVED** [review]

**Comment by manjuransari** [comment]
@selitvin can you please review the request.

**Comment by manjuransari** [comment]
@selitvin can you please review the changes.

**Review by selitvin — APPROVED** [review]
Sorry for the delay - was out for a vacation. Thanks for the contribution.

**Comment by manjuransari** [comment]
@selitvin can you review and approve the running workflows. It says first time contributor need maintainer to approve.

**Comment by selitvin** [comment]
Seems like flake8 fails. Not sure why is that. Will try looking at it today or tomorrow - unless you know what's going on?

**Comment by manjuransari** [comment]
Don't know why the flake8 fails.

**Comment by selitvin** [comment]
Trying to fix flake8 failure here: #701.

File: petastorm/fs_utils.py [hunk 5]
```diff
@@ -134,8 +136,11 @@ def __init__(self, dataset_url, hadoop_configuration=None, connector=HdfsConnect
 
             storage_options = storage_options or {}
             protocol = self._parsed_dataset_url.scheme
-            self._filesystem = fsspec.filesystem(protocol, **storage_options)
-            self._filesystem_factory = lambda: fsspec.filesystem(protocol, **storage_options)
+            cls = fsspec.get_filesystem_class(protocol)
+            options = cls._get_kwargs_from_urls(self._dataset_url)
+            update_storage_options(options, storage_options)
+            self._filesystem = cls(**options)
+            self._filesystem_factory = lambda: cls(**options)
```
Comment (selitvin) [inline petastorm/fs_utils.py:143]:
```suggestion
            self._filesystem_factory = lambda: cls(**options)  # pylint: disable=unnecessary-lambda
```

**Comment by manjuransari** [comment]
@selitvin build/test complete. Waiting for merge.

**Review by chongxiaoc — APPROVED** [review]

**Pull request merged**